win: enable build with vs2017

refack opened the pull request:
Add support for building node with Visual Studio 2017

[Inspired by @bzoz and #11084]
With this version VS does not use Windows registry any more, instead it installs a COM server and a tools called `vswhere` that one can query to obtain list of installed modules and their locations - see this blogpost: https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/ by @heaths

I've written a script that wraps `vswhere` with all the needed arguments for finding a VC++ toolchain, transplanted from [`node4good/msvs-com-helper`](https://github.com/node4good/msvs-com-helper)

Checklist

- [x] make `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines
- [x] update documentation (BUILDING.ms)

Affected core subsystem(s)
win, build, tools

Ref: nodejs/node-gyp#1130, nodejs/node-gyp#1103

/cc @nodejs/build @nodejs/platform-windows

refack commented:
Ref: boostorg/build#170

refack commented:
cc @indutny ;)

mscdex commented:
Are we maintaining gyp now or should the changes in `tools/gyp` be upstreamed somewhere first?

refack commented:
some changes in `GYP` are long time coming (have landed in the last few months). VS2017 specific code is being reviewed upstream https://chromium-review.googlesource.com/c/453201/
(INHO if the code is in the repository it's our responsibility 🤷)

refack commented:
A better version of `GYP` lives in https://github.com/nodejs/node-gyp/tree/master/gyp.
🤞for [`gyp.js`](https://github.com/indutny/gyp.js) to mature

rvagg commented:
> A better version of GYP lives in https://github.com/nodejs/node-gyp/tree/master/gyp.

Should we just copy that version then? Do your changes here diverge it much from the node-gyp version?

@joaocgreis can you comment on this? It looks a bit different to the approach you've taken in node-gyp and I wouldn't mind seeing a unified approach to both.

refack commented:
Rebased on top of  `GYP` version from `node-gyp`

joaocgreis commented:
Took me a few more hours to finish my work on this, but I had it almost done so I completed testing and open a PR: https://github.com/nodejs/node/pull/11867 . We can now compare and use the best or merge.

About this version, I'd like to see the gyp changes in a separate commit. If a user has multiple versions of VS installed and only one with a C++ compiler, does this filter to choose that one?

About the gyp update, I think it's easier to maintain if we update to the upstream version and float patches, as we've been doing. The gyp from node-gyp is missing some floating patches that we have here, at least those would have to be added (compare https://github.com/nodejs/node-gyp/commits/ae141e1/gyp with https://github.com/nodejs/node/commits/1a210c4/tools/gyp). This would be easier to review in a separate PR. (For reference, there's a good guide about floating patches for V8: https://github.com/nodejs/node/blob/master/doc/guides/maintaining-V8.md)

refack commented:
@joaocgreis In Hebrew we say "rivalry among scholars brings much wisdom" 😉

refack commented:
@joaocgreis so I've refactored the C# file, and now the scripts can pass filter criteria to the C# class (such as if if instance has all requirements for VC compilation) and it's easier to get values via `powershell`.
As for how to manage the transplanted `GYP` IMHO a pointer to the commit log in `node-gyp` is enough.

P.S. I'm working on a PR for `node-gyp` as I've did some debugging.

refack commented:
@joaocgreis as for the floating `.patch` files there's nodejs/node-gyp#1122
/cc @bnoordhuis

indutny commented:
What are the major differences between approach here and #11867 ?

refack commented:
I have [3 major reservation](https://github.com/nodejs/node/pull/11867#issuecomment-287228981)
1. __I think that cherry picking changes from `node-gyp/gyp` is gonna be a PITA.__
(so I transplanted the whole directory from `node-gyp`, actualy I think we should use `GYP:HEAD` after [CR456360](https://chromium-review.googlesource.com/c/456360/) lands)
2. __Should at least use my new (v1.10.0) `GetVS2017Configuration.cs` it has gotten a bit of traction, has a simpler interface, and has a few bugs solved.__
(#11867 uses @joaocgreis patch of one of my older `.cs` file)
3. __I think you should run `vcvarsall.bat` in any case.__
(downside, takes a few seconds more in case of `nobuild`. upside; sets all needed ENV variables for the downstream process, needed anyway for the actual `build` action to work)

indutny commented:
Alright, 👍 to this PR then.

seishun commented:
I'd prefer to avoid shipping a hand-rolled JSON stringifier.

indutny commented:
Should we consider using `gyp.js` for building node? :wink:

If I'll have enough time, I will make `gyp.js` work without node.js so it will be usable for bootstraping.

refack commented:
> I'd prefer to avoid shipping a hand-rolled JSON stringifier.

@seishun, me too, but a requirement is to be back-compatible with `.NET 2.0` and `.NET Core` that don't have a JSON stringifier built in :(
Anyway that code is unreachable here and is only there for my own backward compatibility.  Here I use `([VisualStudioConfiguration.Main]::QueryEx($filter))[0]` [directly](https://github.com/nodejs/node/pull/11852/files#diff-f46419730d6c19526bec4a523866b908R12) in `powershell` with no serialization

refack commented:
> Should we consider using gyp.js for building node? 😉
>
> If I'll have enough time, I will make gyp.js work without node.js so it will be usable for bootstraping.

If now `python` is a requirement why not replace that with an previously made `node` binary?

seishun commented:
>Anyway that code is unreachable here and is only there for my own backward compatibility.

Backward compatibility with what?

refack commented:
Code is directly transplanted from [`windows-autoconf`](https://github.com/node4good/windows-autoconf), where I use "JSON" like output as the serialization format from .NET to other environments. So it's not really a JSON serializer, rather an internal helper function...🤔

refack commented:
@seishun you made me think I should make that clear in the commit log

seishun commented:
This C# file is simple enough that I don't see why Node.js couldn't "fork" it and remove all the unnecessary methods.

Alternatively, since you're the maintainer of that repo, maybe you can just separate it into multiple files so node can pull just the necessary code.

refack commented:
🤔
...
...
...
💡
🤓
@seishun good idea! Just got stuck in my own head...

refack commented:
Removed the serialization code from this PR
/cc @seishun

joaocgreis commented on vcbuild.bat:
```diff
@@ -49,7 +48,6 @@ if /i "%1"=="clean"         set target=Clean&goto arg-ok
 if /i "%1"=="ia32"          set target_arch=x86&goto arg-ok
 if /i "%1"=="x86"           set target_arch=x86&goto arg-ok
 if /i "%1"=="x64"           set target_arch=x64&goto arg-ok
-if /i "%1"=="vc2015"        set target_env=vc2015&goto arg-ok
```
This should be kept. `vcbuild.bat` should search for VS2017 and, only if that is not found, VS2015. But a user with both installed should be able to use `vc2015/2017` to specifically select the version to use or fail if that version is not found. This mechanism was introduced in https://github.com/nodejs/node/pull/4645, I don't think it should be removed.

refack replied:
Ok.
It has been a blind switch since [build: remove VS 2013 switch from vcbuild.bat](https://github.com/nodejs/node/commit/a3f861de9d1ab43934fa8de6352a9a98ddefd44a) but now it makes sense again...

joaocgreis commented:
@indutny the main differences between this and https://github.com/nodejs/node/pull/11867 are:
- Filtering VS2017 installations to make sure they are usable to compile C++. This is important because VS2017 adds explicit support for multiple side-by-side installations. @refack is already making some progress here to support this.
- This PR changes `configure` while https://github.com/nodejs/node/pull/11867 uses a small temporary `.bat` file to set environment variables.
- This PR includes a full GYP update (possibly unneeded) while https://github.com/nodejs/node/pull/11867 adds only a small floating patch (submitted upstream).
- This PR ports a lot of code from `windows-autoconf` while https://github.com/nodejs/node/pull/11867 uses only a version trimmed down specifically for this.

It would be nice to see `gyp.js` building node, I believe we're all a bit weary of gyp. In fact, one of the main reasons for dropping https://github.com/nodejs/node/pull/11084 and https://github.com/nodejs/node-gyp/pull/1101 in favor of the C# script was to avoid deepening the python dependency, making it easier for `gyp.js` or something else in the future.

refack commented:
@joaocgreis 

> It's nice to see this getting better! The version filtering is a welcome addition. @refack can you add logging, to make sure users know what is missing from a VS installation that was discarded? Also, please make sure this can still use VS2015 if an unusable installation of VS2017 exists.

Filtering was inspired by your patch 👍, now with all the filtering done in powershell I can easily add logging streamed to `stdout` or `stderr`

>The GYP update, is it needed here for this? Note that nodejs/node-gyp#1122 was only about removing the committed .patch files, not about removing the patches themselves. They can still be listed by git, and I thing that's the easiest way of tracking our changes. There are also some patches in this repository that should not be removed.

This is the version from `node-gyp`. Since we needed an update I thought transplanting that version will keep only one patched `GYP` version to maintain in the whole `nodejs` org (RE nodejs/node-gyp#1151).

>Our CI machines have only VS2015 installed, there are some issues: https://ci.nodejs.org/job/node-compile-windows/7683/label=win-vs2015/console

Looking into it. Also I've added an [appveyor CI](https://ci.appveyor.com/project/refack/node) project.

refack commented:
**"Bring back VS150COMNTOOLS"**
I urge everyone else to up vote at https://developercommunity.visualstudio.com/content/problem/35325/missing-vs150comntools.html

refack commented:
Who usually bumps `npm`? @fishrock123? @MylesBorins?
Getting `node-gyp@3.6.0` would be very nice. Any change of getting `4.5.0` or float patching [package.json#L68](https://github.com/nodejs/node/blob/master/deps/npm/package.json#L68)

gibfahn commented:
It's usually one of the @nodejs/npm team. `npm` is vendored in, so we can't just upgrade the `package.json`.

refack commented:
@gibfahn so what will it take to bump to `v4.5.0`?

refack commented:
my goal is to merge to kill [`tools/gyp`](https://github.com/nodejs/node/tree/master/tools/gyp) and use [`deps/npm/node_modules/node-gyp/gyp`](https://github.com/nodejs/node/tree/master/deps/npm/node_modules/node-gyp/gyp) instead.
Not enough that we use this almost obsolete tool, we have two copies on it in the tree 🤦

__[Long live `gyp.js` ✊]__

MylesBorins commented:
@refack I personally not sure it is the best idea for us to rely on something being shipped in the `npm` tree. Their `node_modules` folder is not part of their support contract and I'd be afraid of that changing on us unexpectedly.

refack commented:
> @refack I personally not sure it is the best idea for us to rely on something being shipped in the npm tree. Their node_modules folder is not part of their support contract and I'd be afraid of that changing on us unexpectedly.

Ok, let's think of a better solution, cause now it's kinda ridiculous. We maintain `node-gyp`, but we get it vendored in through a 3rd party into the tree. Meanwhile we maintain two copies of a 4th party tool (that everyone hates) that we float patches (two different sets) onto... 😵

joaocgreis commented:
@refack can you keep on this PR only the changes necessary to build with VS2017? The Gyp update and the verbose argument should go on their own PRs.

About Gyp, I know we have plenty of room for improvement, but I disagree that what we have is ridiculous. The two copies of Gyp we have are for different purposes and we maintain them differently. The node-gyp version is shipped to all users, floating patches are very discouraged and we should keep it very stable. The version in tools is used only by node developers to build node, we don't need to be as strict with floating patches (and that's actually useful because it may need to do more complex things like cross compiling).

About VS2017, can you trim down the Powershell scripts to have only what we need here? The COM interface will not change and the way we detect VS2017 should not need to change either (there can always be bugs, and it can probably be simplified if if Gyp starts supporting VS2017). Thus, adding the files as they are on `msvs-com-helper` is like adding another dependency, for no benefit.

refack commented:
> @refack can you keep on this PR only the changes necessary to build with VS2017? The Gyp update and the verbose argument should go on their own PRs.

Agreed ss for the `verbose`, the `GYP` update is needed (it's taken from `node-gyp`)
 
> About Gyp, I know we have plenty of room for improvement, but I disagree that what we have is ridiculous. The two copies of Gyp we have are for different purposes and we maintain them differently. The node-gyp version is shipped to all users, floating patches are very discouraged and we should keep it very stable. The version in tools is used only by node developers to build node, we don't need to be as strict with floating patches (and that's actually useful because it may need to do more complex things like cross compiling).

Ok, not ridiculous, but can we agree it's a code smell? I disagree that two versions is a good thing.

> About VS2017, can you trim down the Powershell scripts to have only what we need here? The COM interface will not change and the way we detect VS2017 should not need to change either (there can always be bugs, and it can probably be simplified if if Gyp starts supporting VS2017). Thus, adding the files as they are on msvs-com-helper is like adding another dependency, for no benefit.

The code in `msvs-com-helper` is the whittled down core of `windows-autoconf`, it is receiving review on 4 other projects, hence tweaks and bug fixes, and as such it is a dependency, but in a good way. For as long as I can I'm willing to take responsibility for it, and this way it'll be easier for me to maintain.
tl;dr IMHO a well seasoned tool is better then a custom one.

refack commented:
@joaocgreis done

joaocgreis commented:
> Agreed ss for the `verbose`, the `GYP` update is needed (it's taken from `node-gyp`)

Why is this needed? I did a `git rebase HEAD~3 --onto HEAD~4` and this seems to work just as well.

> Ok, not ridiculous, but can we agree it's a code smell? I disagree that two versions is a good thing.

I think we can agree that all of our patches should land upstream and we would then keep clean versions... but it's quite hard to get a review upstream and we don't want to stop completely.

We can also agree that duplication is a thing that deserves a close look. There is also the way npm distributes node-gyp to consider, as @MylesBorins pointed out. It makes sense for npm to distribute the tool to compile native npm modules (so it always works with the npm version it ships with, gets updated when npm does and so on).

> The code in `msvs-com-helper` is the whittled down core of `windows-autoconf`, it is receiving review on 4 other projects, hence tweaks and bug fixes, and as such it is a dependency, but in a good way. For as long as I can I'm willing to take responsibility for it, and this way it'll be easier for me to maintain.
tl;dr IMHO a well seasoned tool is better then a custom one.

A well seasoned tool is better then a custom one, but not always. This is too simple to justify a well seasoned tool unless it actually makes our code simpler or better. I tried to run this on a VS2017 installation with only the minimal set of components (`Visual Studio C++ core features`, `VC++ 2017 v141 toolset (x86,x64)` and `Windows 10 SDK (10.0.10586.0)`) and it failed with just `Failed to find Visual Studio installation.`. The custom tool from https://github.com/nodejs/node/pull/11867 works, and when it fails it displays exactly why, so I could argue that the custom tool is the well seasoned one. Was there any review feedback from any of those 4 other projects that would have made this better here? Can you give an example where `msvs-com-helper` is better for node here than the custom tool?

refack commented:
> Why is this needed? I did a git rebase HEAD~3 --onto HEAD~4 and this seems to work just as well.

current version does not set `<WindowsTargetPlatformVersion>10.0.14393.0</WindowsTargetPlatformVersion>` so it works only is SDK8.1 is installed.

> A well seasoned tool is better then a custom one, but not always. This is too simple to justify a well seasoned tool unless it actually makes our code simpler or better. I tried to run this on a VS2017 installation with only the minimal set of components (Visual Studio C++ core features, VC++ 2017 v141 toolset (x86,x64) and Windows 10 SDK (10.0.10586.0)) and it failed with just Failed to find Visual Studio installation.. The custom tool from #11867 works, and when it fails it displays exactly why, so I could argue that the custom tool is the well seasoned one. Was there any review feedback from any of those 4 other projects that would have made this better here? Can you give an example where msvs-com-helper is better for node here than the custom tool?

My main argument is that since I'm willing to maintain this tool, it is safer and more robust for me to reuse the tool as is rather than customize for node. And since it's me, node has a highway to the "upstream". I'm willing to improve the tool to node's requirements.
Some improvements the tool has received that are beneficial to node:
1. Separation of COM marshaling (`.cs`) from and filtering and projection logic (`.ps1`)
2. Removed JSON serialization in favor of direct .NET/Powershell type sharing
3. More precise filtering of suitable VS2017 instance
4. Support for multiple VS2017 installations
5. Support for the Buildtools SKU (not only the Community/Pro/Enterprise SKUs) 
6. Made compatible with [PowershellCore](https://github.com/PowerShell/PowerShell)

We can get more verbose response if we add `-Information` to the powershell call, but I moved it into the `verbose` patch.
If you think the error message is lacking, tell me, or create a PR.

refack commented on vcbuild.bat:
```diff
@@ -165,7 +190,8 @@ set PLATFORM_TOOLSET=v140
 :msbuild-not-found
-echo Failed to find Visual Studio installation.
+echo Failed to find a suitable Visual Studio installation.
+Echo Please consult https://github.com/nodejs/node/blob/master/BUILDING.md#windows-1
```
@joaocgreis Added a better error message

refack commented:
7. The full tools has diagnostic capabilities to help users with problems
    ` powershell -File Tools\msvs-com-helper\GetKey.ps1 * *`
8. If #12310 is completed, then we'll have a perfect fit

refack commented:
Great news we landed VS2017 support to `GYP` https://chromium-review.googlesource.com/c/453201/

AndrewPardoe commented:
The officially supported vswhere tool is now available as part of the install starting today with Visual Studio 15.2 preview 2: https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/. I hope this helps your scenario. 

This doesn't fix all the existing installs of VS 2017 but going forward you can now detect the VS install locations and whether C++ tools are available from a scripted environment.

refack commented:
Updated to use `vswhere`
@joaocgreis PTAL
/cc @nodejs/platform-windows

joaocgreis commented:
I'm +1 for using `vswhere` here, we can simply not support early versions of VS2017 that did not have it. (Note that the same is not true for node-gyp, for users we should reach for maximum compatibility.)

@refack can you make sure this PR only has changes that relate to the objective here? The changes to `configure` and `gyp_node.py` look like a nice simplification, but would be easier to review in their own PR (since that affects all platforms we could bring other people that are not concerned with VS2017). Also, can you keep the parameters in vcbuild named `vc201x`? If you feel strongly about it, please open another PR. The change is good for consistency, but there might be build farms depending on what's there now so I'd like to see it land on its own with longer notice.

refack commented:
@joaocgreis PTAL. If you approve I'll land.

bzoz commented on vcbuild.bat:
```diff
@@ -51,7 +51,8 @@ if /i "%1"=="clean"         set target=Clean&goto arg-ok
 if /i "%1"=="x64"           set target_arch=x64&goto arg-ok
-if /i "%1"=="vc2015"        set target_env=vc2015&goto arg-ok
+if /i "%1"=="vs2015"        set target_env=vs2015&goto arg-ok
+if /i "%1"=="vs2017"        set target_env=vs2017&goto arg-ok
```
Those should be `vc2015` and `vc2017`

refack replied:
I won't bikeshed over this if you feel strongly.
But a lllooooonnngggg and boring discussion on the boost mailing list came to a conclusion it should be `vs` since vc2015 or vc2017 are misnomers, the VC++ versions are either 14.0 and 15.0 or v140 and v141.
Ref: https://lists.boost.org/Archives/boost/2017/03/233597.php

bzoz replied:
People right now can have setups that call `vcbuild vc2015`, no reason to break those.

refack replied:
Good enough

refack replied:
I'll add it as an "undocumented" switch, what do you think:
```cmd
if /i "%1"=="vc2015"        set target_env=vs2015&goto arg-ok
```

bzoz commented on vcbuild.bat:
```diff
@@ -144,7 +145,33 @@ if defined noprojgen if defined nobuild if not defined sign if not defined msi g
 
 @rem Set environment for msbuild
 
+set msvs_host_arch=x86
```
Shouldn't this go inside `:vs-set-2017`? It looks like it is only used there.

refack replied:
It actually should be added to the VS2015 case as well, but that's for a different PR

refack replied:
Also it's kinda consistent with the `rem` two lines above ;)

refack replied:
I'm testing the implication on calling the VS2016 with a `host_target` arch

bzoz commented on vcbuild.bat:
```diff
@@ -144,7 +145,33 @@ if defined noprojgen if defined nobuild if not defined sign if not defined msi g
+call tools\msvs\vswhere_usability_wrapper.cmd
+if "_%VCINSTALLDIR%_" == "__" goto vs-set-2015
+set vcvars_call="%VCINSTALLDIR%\Auxiliary\Build\vcvarsall.bat" %vcvarsall_arg%
+echo %vcvars_call%
```
I dont think this is needed.

refack replied:
The `ECHO`?
It's used in:
The call to `configure` https://github.com/refack/node/blob/3d416125b23827f76eae2c5e69a84a9b12557075/vcbuild.bat#L213
The call to `cctest` https://github.com/refack/node/blob/3d416125b23827f76eae2c5e69a84a9b12557075/vcbuild.bat#L403
Personally I find it very helpful, but it could move to a "verbosing" PR

refack commented:
/cc @nodejs/build @nodejs/platform-windows
Updated PR to use `vswhere`

refack commented:
ping @nodejs/build @nodejs/platform-windows @joaocgreis @bzoz @indutny 
Comments have been addressed PTAL 🙏

gibfahn commented:
LGTM as long as @seishun @bzoz and @joaocgreis are okay with it.

joaocgreis commented:
CI: https://ci.nodejs.org/view/All/job/node-test-pull-request/7858/

joaocgreis commented on vcbuild.bat:
```diff
@@ -144,7 +147,33 @@ if defined noprojgen if defined nobuild if not defined sign if not defined msi g
+@rem usualy vcvarsall takes an argument: host + '_' + target
+set vcvarsall_arg=%msvs_host_arch%_%target_arch%
+@rem unless both host and taget are x64
+if %target_arch%==x64 if %msvs_host_arch%==amd64 set vcvarsall_arg=amd64
```
You could [`use VsDevCmd`](https://github.com/janeasystems/node/blob/d085fcf95fb5d248b2de46a936c0ea61be554147/vcbuild.bat#L157-L160) to avoid having to detect the host arch.

refack replied:
I kinda don't trust `VsDevCmd` 🤷‍♀️ I think it does too much.
Also both of them, for x86 it choose a native compiler over a 64bit host cross compiler:
```
c:\code$ "c:\bin\dev\VS\Microsoft Visual Studio 2017 Community\Common7\Tools\VsDevCmd.bat" -arch=x86
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0.26403.7
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************

c:\code$ where cl
c:\bin\dev\VS\Microsoft Visual Studio 2017 Community\VC\Tools\MSVC\14.10.25017\bin\HostX86\x86\cl.exe
```
And apparently that matters: https://chromium-review.googlesource.com/c/486400/#message-04d641d86728a1b1832181033b692eb9a85e55de

refack replied:
*"does too much" == "runs longer" (probably does .NET setup)

refack commented:
@bzoz PTAL, vc2015 is retained as an "undocumented" compatibility switch

refack commented:
@joaocgreis @gibfahn are we waiting for @bzoz? I've addressed his comments...
(this is kind of a PITA for me since I need to float over to all my other branches, and be careful not to leak it there...)

refack commented:
New CI: https://ci.nodejs.org/job/node-test-commit/9646/

gibfahn commented:
Landed in https://github.com/nodejs/node/commit/1c93e8c94b5d0c304e20b2af3050db0bdc8864d0

refack commented:
> Landed in 1c93e8c

Thanks I didn't even notice...
